Redundant workgroup metadata CI check

scholtzan opened the pull request:
## Description
Follow-up to https://github.com/mozilla/bigquery-etl/pull/6811#issuecomment-2593766594

This adds a new CI check to ensure that there are no table-level access definitions that are redundant with dataset access definitions. This PR also removes configs that are redundant

<!--
Please reference related Jira tickets, GitHub issues or Bugzilla. This repo has been
configured to automatically insert hyperlinks for DSRE and DENG tickets.
See https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/configuring-autolinks-to-reference-external-resources
-->

**Reviewer, please follow [this checklist](https://github.com/mozilla/bigquery-etl/blob/main/.github/reviewer_checklist.md)**

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/DENG-7402)

whd reviewed (approved):
Diff LGTM, thanks for making this more ergonomic

whd commented:
One more thing it would be great to add as a CI check: I think we should remove `default_table_workgroup_access` from `main` and require it not be set via a CI check. It should always have the same value as `workgroup_access` and is used exclusively by metadata generation logic to decide whether to set `workgroup_access` differently in generated metadata.

scholtzan commented on .circleci/workflows.yml:
```diff
@@ -289,6 +289,10 @@ jobs:
             - checkout
             - *restore_venv_cache
             - *build
+            - run:
+                name: Validate workgroup access configuration on main
+                command: |
+                  PATH="venv/bin:$PATH" script/bqetl metadata validate-workgroups sql/
```
Run the metadata workgroup access validation before generated SQL gets copied over. During SQL generation also metadata is generated, which updates and sets the `default_table_workgroup_access` and `workgroup_access` configs and would make the CI check fail.

scholtzan commented on bigquery_etl/metadata/validate_metadata.py:
```diff
@@ -284,6 +284,73 @@ def validate_exclusion_list_expiration_days(metadata, path):
+    return is_valid
+
+
+def validate_default_table_workgroup_access(path):
+    """
+    Check that default_table_workgroup_access does not exist in metadata.
+
+    default_table_workgroup_access will be generated from workgroup_access and
+    should not be overridden.
+    """
+    is_valid = True
+    dataset_metadata_path = Path(path).parent.parent / "dataset_metadata.yaml"
+    if not dataset_metadata_path.exists():
+        return is_valid
+
+    with open(dataset_metadata_path, "r") as yaml_stream:
+        try:
+            metadata = yaml.safe_load(yaml_stream)
+        except yaml.YAMLError as e:
+            raise e
+
+    if "default_table_workgroup_access" in metadata:
+        is_valid = False
+        click.echo(
+            click.style(
+                f"ERROR: default_table_workgroup_access should not be explicity specified in {dataset_metadata_path}. "
+                + "The default_table_workgroup_access configuration will be automatically generated.",
+                fg="red",
+            )
+        )
+
+    return is_valid
```
Added an additional check to make sure `default_table_workgroup_access` is not being set explicitly. This check runs on non-generated files (which should be equal to files that are on `main`).

scholtzan commented:
@BenWu @whd Might be good if you can take another look at this. I added a check as described in https://github.com/mozilla/bigquery-etl/pull/6822#issuecomment-2596474076 and removed the `default_table_workgroup_access` from `main`

BenWu reviewed (commented):
Shouldn't the `default_table_workgroup_access` still be in the sql.diff because it's generated?  I think the metadata update might need some changes because `default_table_workgroup_access` is never null https://github.com/mozilla/bigquery-etl/blob/3282192a21972f99e9bd9bb0fb98570bfecc07f7/bigquery_etl/metadata/parse_metadata.py#L493-L496

BenWu commented:
I triggered CI and then cancelled it so it's showing up as failed.  I'll just rerun it
